GSA/digitalgov.gov
# Static file upload type for S3

### Summary
Allow static file uploading for S3 hosting to avoid storage space issues and to create one place for all files to be stored.

### Preview Link
[Preview](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-workflow-file-upload/workflow/#/collections/file-upload)

#5670 is a test upload using the new collection type

### Current process

1. Upload img/file via workflow and publish
2. circleci runs `gulp img` that does some processing for images and uploads to s3
3. cloudpages build runs

### Steps to Recreate

1. Visit [workflow](https://federalist-466b7d92-5da1-4208-974f-d61fd4348571.app.cloud.gov/preview/gsa/digitalgov.gov/nl-workflow-file-upload/workflow/#/collections/file-upload)
2. Click on "Upload File" button
3. Upload static file (.pdf, .ppt/.pptm/.pptx, .xlsx/.xls, .doc) and Publish
4. Check S3 bucket now contains the new file

### Testing Locally

1. copy file to `content/uploads/_inbox`
2. run `gulp upload` to process images/files or if running `npm run start` the upload process will automatically run
3. verify new file is saved to s3 bucket


### Upload Scripts
1. upload image/file in workflow
2. `file-prep` creates working directories, normalizes file name and generates yml data file
3. `file-process` creates responsive images variants
4. `file-upload` looks in image/file directories respectively and uploads to their s3 buckets


### File Directories

These directories are generated and removed when `gulp upload` is executed on circleci:

- `content/uploads/_inbox/` location of original upload location
- `content/uploads/_working-images/to-process/`: image file with normalized name
- `content/uploads/_working-images/processed/`: responsive variants are stored here, uploaded to s3
- `content/uploads/_working-files/to-process/`: static file with normalized name, uploaded to s3

<img width="1071" alt="Screen Shot 2022-12-16 at 2 52 27 PM" src="https://user-images.githubusercontent.com/104778659/208178177-4080eebf-e61e-4c65-8cd8-9e56e86a84f3.png">

### Errors

Incorrect file types will display an error when uploading wrong format:

<img width="1680" alt="Screen Shot 2022-12-19 at 10 37 17 AM" src="https://user-images.githubusercontent.com/104778659/208463338-0dbf94c6-44c7-4eb6-b701-98e10409e56b.png">

### Build Issues

It looks like circleci is still using the `.circleci/config.yml` from main and not the test branch during the build.
Images are not uploaded due to running the old upload scripts which references `content/images/**` and not the new location `content/uploads/**`

Should be seeing `gulp upload` when executing, instead it calls `gulp img`.

<img width="1380" alt="Screen Shot 2023-01-20 at 9 59 16 AM" src="https://user-images.githubusercontent.com/104778659/213733823-abb76615-8a81-4421-8e25-2deabff4a43a.png">
